File: app/src/main/java/com/hyf/intelligence/kotlin/utils/TimeUtils.java
package com.hyf.intelligence.kotlin.utils;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.TimeZone;
import java.util.Date;
import java.util.Calendar;

public class TimeUtils {
    public static String[] getFormatedTime(long time) {
        SimpleDateFormat formatter = new SimpleDateFormat("HH:mm:ss");
        formatter.setTimeZone(TimeZone.getTimeZone("GMT"));
        String hms = formatter.format(time);
        return hms.split(":");
    }

    public static String AddTime(String begain, int end) {
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
        Date date = null;
        try {
            date = sdf.parse(begain);
        } catch (ParseException e) {
            e.printStackTrace();
        }
        Calendar ca = Calendar.getInstance();
        ca.setTime(date);
        ca.add(Calendar.HOUR_OF_DAY, end);
        System.out.println(sdf.format(ca.getTime()));
        return sdf.format(ca.getTime());
    }

    public static String gettime() {
        Date date = new Date();
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm");
        String dateString = sdf.format(date);
        return dateString;
    }
    public static String convert(String time) {
        SimpleDateFormat sdr = new SimpleDateFormat("yyyy-MM-dd HH:mm");
        @SuppressWarnings("unused")
        long lcc = Long.valueOf(time);
        int i = Integer.parseInt(time);
        String times = sdr.format(new Date(i * 1000L));
        return times;

    }

    public static Long dateDiff(String startTime, String endTime, String str) {
        // 按照传入的格式生成一个simpledateformate对象
        SimpleDateFormat sd = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
        long nd = 1000 * 24 * 60 * 60;// 一天的毫秒数
        long nh = 1000 * 60 * 60;// 一小时的毫秒数
        long nm = 1000 * 60;// 一分钟的毫秒数
        long ns = 1000;// 一秒钟的毫秒数
        long diff;
        long day = 0;
        long hour = 0;
        long min = 0;
        long sec = 0;

        float diffs;
        float nhs = 1000 * 60 * 60;// 一小时的毫秒数
        float hours = 0;
        // 获得两个时间的毫秒时间差异
        try {
            diff = sd.parse(endTime).getTime() - sd.parse(startTime).getTime();
            diffs = sd.parse(endTime).getTime() - sd.parse(startTime).getTime();
            day = diff / nd;// 计算差多少天
            hour = diff % nd / nh + day * 24;// 计算差多少小时
            min = diff % nd % nh / nm + day * 24 * 60;// 计算差多少分钟
            sec = diff % nd % nh % nm / ns;// 计算差多少秒
            hours = diffs / nhs;
            // 输出结果
            System.out.println("时间相差：" + day + "天" + (hour - day * 24) + "小时"
                    + (min - day * 24 * 60) + "分钟" + sec + "秒。");
            System.out.println("diff="+diff+",hours=" + hour + ",min=" + min);
            if (str.equalsIgnoreCase("h")) {
                return hour;
            } else {
                return min;
            }

        } catch (ParseException e) {
            e.printStackTrace();
        }
        if (str.equalsIgnoreCase("h")) {
            return hour;
        } else {
            return min;
        }
    }

    public static String dateDiffs(String startTime, String endTime, String str) {
        // 按照传入的格式生成一个simpledateformate对象
        SimpleDateFormat sd = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
        float nd = 1000 * 24 * 60 * 60;// 一天的毫秒数
        float nh = 1000 * 60 * 60;// 一小时的毫秒数
        float nm = 1000 * 60;// 一分钟的毫秒数
        float diff;
        float day = 0;
        float hour = 0;
        float min = 0;
        // 获得两个时间的毫秒时间差异
        try {
            diff = sd.parse(endTime).getTime() - sd.parse(startTime).getTime();
            day = diff / nd;// 计算差多少天
            hour = diff % nd / nh + day * 24;// 计算差多少小时
            min = diff % nd % nh / nm + day * 24 * 60;// 计算差多少分钟
            if (str.equalsIgnoreCase("h")) {
                return String.valueOf(hour);
            } else {
                return String.valueOf(min);
            }

        } catch (ParseException e) {
            e.printStackTrace();
        }
        if (str.equalsIgnoreCase("h")) {
            return String.valueOf(hour);
        } else {
            return String.valueOf(min);
        }
    }

    public static String getDateHours(String time, String str) {
        // 按照传入的格式生成一个simpledateformate对象
        SimpleDateFormat sd = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
        float nd = 1000 * 24 * 60 * 60;// 一天的毫秒数
        float nh = 1000 * 60 * 60;// 一小时的毫秒数
        float nm = 1000 * 60;// 一分钟的毫秒数
        float diff;
        float day = 0;
        float hour = 0;
        float min = 0;
        // 获得两个时间的毫秒时间差异
        try {
            diff = sd.parse(time).getTime();
            day = diff / nd;// 计算差多少天
            hour = diff % nd / nh + day * 24;// 计算差多少小时
            min = diff % nd % nh / nm + day * 24 * 60;// 计算差多少分钟
            if (str.equalsIgnoreCase("h")) {
                return String.valueOf(hour);
            } else {
                return String.valueOf(min);
            }

        } catch (ParseException e) {
            e.printStackTrace();
        }
        if (str.equalsIgnoreCase("h")) {
            return String.valueOf(hour);
        } else {
            return String.valueOf(min);
        }
    }
}

File: app/src/main/java/com/hyf/intelligence/kotlin/utils/TimeUtils.kt
package com.hyf.intelligence.kotlin.utils

import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.TimeZone
import java.util.Date
import java.util.Calendar

object TimeUtils {
    fun getFormatedTime(time: Long): Array<String> {
        val formatter = SimpleDateFormat("HH:mm:ss")
        formatter.timeZone = TimeZone.getTimeZone("GMT")
        val hms = formatter.format(time)
        return hms.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
    }

    fun AddTime(begain: String, end: Int): String {
        val sdf = SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
        var date: Date? = null
        try {
            date = sdf.parse(begain)
        } catch (e: ParseException) {
            e.printStackTrace()
        }

        val ca = Calendar.getInstance()
        ca.time = date
        ca.add(Calendar.HOUR_OF_DAY, end)
        println(sdf.format(ca.time))
        return sdf.format(ca.time)
    }

    fun gettime(): String {
        val date = Date()
        val sdf = SimpleDateFormat("yyyy-MM-dd HH:mm")
        return sdf.format(date)
    }

    fun convert(time: String): String {
        val sdr = SimpleDateFormat("yyyy-MM-dd HH:mm")
        val lcc = java.lang.Long.valueOf(time)
        val i = Integer.parseInt(time)
        return sdr.format(Date(i * 1000L))

    }

    fun dateDiff(startTime: String, endTime: String, str: String): Long? {
        // 按照传入的格式生成一个simpledateformate对象
        val sd = SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
        val nd = (1000 * 24 * 60 * 60).toLong()// 一天的毫秒数
        val nh = (1000 * 60 * 60).toLong()// 一小时的毫秒数
        val nm = (1000 * 60).toLong()// 一分钟的毫秒数
        val ns: Long = 1000// 一秒钟的毫秒数
        val diff: Long
        var day: Long = 0
        var hour: Long = 0
        var min: Long = 0
        var sec: Long = 0

        val diffs: Float
        val nhs = (1000 * 60 * 60).toFloat()// 一小时的毫秒数
        var hours = 0f
        // 获得两个时间的毫秒时间差异
        try {
            diff = sd.parse(endTime).time - sd.parse(startTime).time
            diffs = (sd.parse(endTime).time - sd.parse(startTime).time).toFloat()
            day = diff / nd// 计算差多少天
            hour = diff % nd / nh + day * 24// 计算差多少小时
            min = diff % nd % nh / nm + day * 24 * 60// 计算差多少分钟
            sec = diff % nd % nh % nm / ns// 计算差多少秒
            hours = diffs / nhs
            // 输出结果
            println("时间相差：" + day + "天" + (hour - day * 24) + "小时"
                    + (min - day * 24 * 60) + "分钟" + sec + "秒。")
            println("diff=$diff,hours=$hour,min=$min")
            return if (str.equals("h", ignoreCase = true)) {
                hour
            } else {
                min
            }

        } catch (e: ParseException) {
            e.printStackTrace()
        }

        return if (str.equals("h", ignoreCase = true)) {
            hour
        } else {
            min
        }
    }

    fun dateDiffs(startTime: String, endTime: String, str: String): String {
        // 按照传入的格式生成一个simpledateformate对象
        val sd = SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
        val nd = (1000 * 24 * 60 * 60).toFloat()// 一天的毫秒数
        val nh = (1000 * 60 * 60).toFloat()// 一小时的毫秒数
        val nm = (1000 * 60).toFloat()// 一分钟的毫秒数
        val diff: Float
        var day = 0f
        var hour = 0f
        var min = 0f
        // 获得两个时间的毫秒时间差异
        try {
            diff = (sd.parse(endTime).time - sd.parse(startTime).time).toFloat()
            day = diff / nd// 计算差多少天
            hour = diff % nd / nh + day * 24// 计算差多少小时
            min = diff % nd % nh / nm + day * 24f * 60f// 计算差多少分钟
            return if (str.equals("h", ignoreCase = true)) {
                hour.toString()
            } else {
                min.toString()
            }

        } catch (e: ParseException) {
            e.printStackTrace()
        }

        return if (str.equals("h", ignoreCase = true)) {
            hour.toString()
        } else {
            min.toString()
        }
    }

    fun getDateHours(time: String, str: String): String {
        // 按照传入的格式生成一个simpledateformate对象
        val sd = SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
        val nd = (1000 * 24 * 60 * 60).toFloat()// 一天的毫秒数
        val nh = (1000 * 60 * 60).toFloat()// 一小时的毫秒数
        val nm = (1000 * 60).toFloat()// 一分钟的毫秒数
        val diff: Float
        var day = 0f
        var hour = 0f
        var min = 0f
        // 获得两个时间的毫秒时间差异
        try {
            diff = sd.parse(time).time.toFloat()
            day = diff / nd// 计算差多少天
            hour = diff % nd / nh + day * 24// 计算差多少小时
            min = diff % nd % nh / nm + day * 24f * 60f// 计算差多少分钟
            return if (str.equals("h", ignoreCase = true)) {
                hour.toString()
            } else {
                min.toString()
            }

        } catch (e: ParseException) {
            e.printStackTrace()
        }

        return if (str.equals("h", ignoreCase = true)) {
            hour.toString()
        } else {
            min.toString()
        }
    }
}
